specs/eventing/test-plan/broker-lifecycle-conformance.md

```diff
@@ -0,0 +1,188 @@
+```
+
+
+## [Test] Immutability
+
+Check for default annotations, this should return the name of the selected implementation: 
+
+```
+kubectl get broker conformance-broker -o jsonpath='{.metadata.annotations.eventing\.knative\.dev/broker\.class}'
+```
+
+Try to patch the annotation: `eventing.knative.dev/broker.class` to see if the resource mutates: 
+
+```
+kubectl patch broker conformance-broker --type merge -p '{"metadata":{"annotations":{"eventing.knative.dev/broker.class":"mutable"}}}'
+```
+
+You should get the following error: 
+```
+Error from server (BadRequest): admission webhook "validation.webhook.eventing.knative.dev" denied the request: validation failed: Immutable fields changed (-old +new): annotations
+{string}:
+	-: "MTChannelBasedBroker"
+	+: "mutable"
+```
+
+### [Output]
+
+```
+{
+  "test": "control-plane/broker-lifecycle/immutability-1"
+  "output": {
+    	"brokerImplementation": "<BROKER IMPLEMENTATION>",
+	"expectedError": "<EXPECTED ERROR>"
+  }
+}
+```
+
+Try to mutate the `.spec.config` to see if the resource mutates: 
+
+```
+kubectl patch broker conformance-broker --type merge -p '{"spec":{"config":{"apiVersion":"v1"}}}'
+```
+
+
+### [Output]
+
+```
+{
+  "test": "control-plane/broker-lifecycle/immutability-2"
+  "output": {
+  	"brokerImplementation": "<BROKER IMPLEMENTATION>",
+	"expectedError": "<EXPECTED ERROR>"
+  }
+}
+```
+
+
+## [Test] Broker Readyness 
+
+Check for condition type `Ready` with status `True`: 
+
+```
+ kubectl get broker conformance-broker -ojson | jq '.status.conditions[] |select(.type == "Ready")' 
+```
+
+### [Output]
+
+```
+{
+  "test": "control-plane/broker-lifecycle/broker-readyness"
+  "output": {
+  	"brokerImplementation": "<BROKER IMPLEMENTATION>",
+	"expectedType": "Ready",
+	"expectedStatus": "True"
+  }
+}
+```
+
+## [Test] Broker is Addresable
+
+Running the following command should return a URL
+
+```
+kubectl get broker conformance-broker -ojson | jq .status.address.url
+```
+
+### [Output]
+
+```
+{
+  "test": "control-plane/broker-lifecycle/broker-addressable"
+  "output": {
+  	"brokerImplementation": "",
+	"obtainedURL": "<BROKER URL>",
+  }
+}
+```
+
+## [Pre] Create Trigger for Broker
+
+Create a trigger that points to the broker:
+
+```
+kubectl apply -f control-plane/broker-lifecycle/trigger.yaml
+```
+
+## [Test] Broker Reference in Trigger
+
+Check that the `Trigger` is making a reference to the `Broker`, this should return the name of the broker.
+
+```
+kubectl get trigger conformance-trigger -ojson | jq '.spec.broker'
+```
+
+### [Output]
+
+```
+{
+  "test": "control-plane/broker-lifecycle/broker-reference-in-trigger"
+  "output": {
+  	"brokerImplementation": "<BROKER IMPLEMENTATION>",
+	"expectedReference": "conformance-broker"
+  }
+}
+```
+
+## [Test] Trigger for Broker Readyness
+
+Check for condition type `Ready` with status `True`: 
+
+```
+kubectl get trigger conformance-trigger -ojson | jq '.status.conditions[] |select(.type == "Ready")'
```

Comment:
instead of `jq`, what about using `-ojsonpath` ? The benefit is to remove the dependency on `jq`

Reply:
yes.. I think that is a good idea.. can you propose the changes?

Reply:
`kubectl get trigger conformance-trigger -ojsonpath="{.status.conditions[?(@.type == \"Ready\")].status}"`

Output:

`True`

Reply:
@lionelvillard yeah.. I got convinced now.. let's remove `jq` from here :) I will do the other changes early on Monday.

Reply:
@lionelvillard `jq` removed now.